libraries/AP_HAL_PX4/CAN.cpp

```diff
@@ -0,0 +1,1035 @@
+        } else {
+            frame.id = uavcan::CanFrame::MaskExtID & (rf.RIR >> 3);
+            frame.id |= uavcan::CanFrame::FlagEFF;
+        }
+
+        if ((rf.RIR & bxcan::RIR_RTR) != 0) {
+            frame.id |= uavcan::CanFrame::FlagRTR;
+        }
+
+        frame.dlc = rf.RDTR & 15;
+
+        frame.data[0] = uint8_t(0xFF & (rf.RDLR >> 0));
+        frame.data[1] = uint8_t(0xFF & (rf.RDLR >> 8));
+        frame.data[2] = uint8_t(0xFF & (rf.RDLR >> 16));
+        frame.data[3] = uint8_t(0xFF & (rf.RDLR >> 24));
+        frame.data[4] = uint8_t(0xFF & (rf.RDHR >> 0));
+        frame.data[5] = uint8_t(0xFF & (rf.RDHR >> 8));
+        frame.data[6] = uint8_t(0xFF & (rf.RDHR >> 16));
+        frame.data[7] = uint8_t(0xFF & (rf.RDHR >> 24));
+
+        *rfr_reg = bxcan::RFR_RFOM | bxcan::RFR_FOVR | bxcan::RFR_FULL; // Release FIFO entry we just read
+
+        /*
+         * Store with timeout into the FIFO buffer and signal update event
+         */
+        CanRxItem frm;
+        frm.frame = frame;
+        frm.flags = 0;
+        frm.utc_usec = utc_usec;
+        rx_queue_.push(frm);
+
+        had_activity_ = true;
+        update_event_.signalFromInterrupt();
+
+        pollErrorFlagsFromISR();
+    }
+}
+
+void PX4CAN::pollErrorFlagsFromISR()
+{
+    const uint8_t lec = uint8_t((can_->ESR & bxcan::ESR_LEC_MASK) >> bxcan::ESR_LEC_SHIFT);
+    if (lec != 0) {
+        can_->ESR = 0;
+        error_cnt_++;
+
+        // Serving abort requests
+        for (int i = 0; i < NumTxMailboxes; i++) { // Dear compiler, may I suggest you to unroll this loop please.
+            TxItem& txi = pending_tx_[i];
+            if (txi.pending && txi.abort_on_error) {
+                can_->TSR = TSR_ABRQx[i];
+                txi.pending = false;
+                served_aborts_cnt_++;
+            }
+        }
+    }
+}
+
+void PX4CAN::discardTimedOutTxMailboxes(uavcan::MonotonicTime current_time)
+{
+    CriticalSectionLocker lock;
+    for (int i = 0; i < NumTxMailboxes; i++) {
+        TxItem& txi = pending_tx_[i];
+        if (txi.pending && txi.deadline < current_time) {
+            can_->TSR = TSR_ABRQx[i];  // Goodnight sweet transmission
+            txi.pending = false;
+            error_cnt_++;
+        }
+    }
+}
+
+bool PX4CAN::canAcceptNewTxFrame(const uavcan::CanFrame& frame) const
+{
+    /*
+     * We can accept more frames only if the following conditions are satisfied:
+     *  - There is at least one TX mailbox free (obvious enough);
+     *  - The priority of the new frame is higher than priority of all TX mailboxes.
+     */
+    {
+        static const uint32_t TME = bxcan::TSR_TME0 | bxcan::TSR_TME1 | bxcan::TSR_TME2;
+        const uint32_t tme = can_->TSR & TME;
+
+        if (tme == TME) {   // All TX mailboxes are free (as in freedom).
+            return true;
+        }
+
+        if (tme == 0) {     // All TX mailboxes are busy transmitting.
+            return false;
+        }
+    }
```

Comment:
This has access to MCU's registers and apart from scope limiting, I see possible threat of optimizing compiler misbehavior if scope brackets are removed.